---
title: "Biomarker Discovery in Immunotherapy Response"
author:
- name: Farnoosh Abbas-Aghababazadeh
  affiliation:
  - &pm Bioinformatics and Computational Genomics Laboratory, Princess Margaret Cancer Center,University Health Network, Toronto, Ontario, Canada
---

# Loading Functions

First we load the functions into the workspace. To compute gene associations with immunotherapy response, calculate signature scores, assess signature associations with immunotherapy response, integrate the estimates, and compare them, we need to load all the necessary packages

```{r load, echo=TRUE, message=FALSE, warning=FALSE}

library(stringr)
app_dir <- str_split(rstudioapi::getActiveDocumentContext()$path,'PredictioR')[[1]][1]

source(file.path(app_dir, 'PredictioR_backup', "R/getHR.R"))
source(file.path(app_dir, 'PredictioR_backup', "R/getSummarizedExperiment.R"))
source(file.path(app_dir, 'PredictioR_backup', "R/getGeneAssociation.R"))
source(file.path(app_dir, 'PredictioR_backup', "R/getGeneSigAssociation.R"))
source(file.path(app_dir, 'PredictioR_backup', "R/getMetaAnalysis.R"))
source(file.path(app_dir, 'PredictioR_backup', "R/getGeneSigScore.R"))

library(survcomp)
library(GSVA)
library(dplyr)
library(meta)
library(metafor)
library(forestplot)
library(ggplot2)
library(ggrepel)
library(gridExtra)
library(data.table)
library(kableExtra)
library(summarytools)
library(MultiAssayExperiment)

```


# Load Immunotherapy Datasets 

The following clinical multimodal immunotherapy datasets are publicly available on [GitHub](https://github.com/bhklab/ClinSets.git) data repository. Datasets are also obtained from the a cloud-based [ORCESTRA](https://www.orcestra.ca/clinical_icb) platform to ensure that the curation and processing of the clinical genomic data are fully transparent and reproducible. 

The immunotherapy dataset is stored in a SummarizedExperiment object. For RNA profiles, we utilize log2-transformed TPM (Transcripts Per Million) data from protein-coding genes. For each dataset, genes with zero expression value in at least 50\% of the samples are filtered out. Note that only studies with at least 15 patients are included in the following analyses.

As an example, the Bladder (Mariathasan,[PMID 29443960](https://pubmed.ncbi.nlm.nih.gov/31792460/)) includes RNA expression, clinical characteristics, and gene meta data. 195 patients have both clinical data and RNA expression data. In addition, this dataset has 17,993 protein-coding genes. 

The detailed clinical characteristics of the Mariathasan dataset including cancer type, age, sex, response (R vs. NR), overall survival (OS), and progression-free survival (PFS). 

```{r load braun data}

# gene expression, clinical characteristics, and gene meta data 

# TODO: dir directory make sense to me here 
dir <- file.path(app_dir, 'PredictioR_backup', "recent_data/se objects/") 


# gene expression, clinical characteristics, and gene meta data 
load(file.path(dir, "ICB_Ravi__Lung__PD-L1.rda") )

#"~/BHK lab//PredictioR/recent_data/seobjects/ICB_IMmotion150__Kidney__IO+targeted.rda"
#"~/BHK lab//PredictioR/recent_data/se objects/ICB_IMmotion150__Kidney__PD-(L)1.rda"

expr <- assay(dat_icb)
clin <- colData(dat_icb) %>% as.data.frame()
#TODO: added for WOLF
clin$response <- factor(clin$response, labels = c("NR", "R"))
annot <- rowData(dat_icb) %>% as.data.frame()

# summary table of selected clinical variable 
sub_clin <- clin[, c("age","sex", "response","survival_time_os", "survival_time_pfs")]
print(dfSummary(sub_clin, 
                style = "grid", 
                plain.ascii = FALSE, 
                graph.col = FALSE), 
      method = 'render')

```

# Biomarkers and Immunotherapy Response Association

Association of specific biomarkers with immunotherapy response (R vs NR) is assessed using a logistic regression model. Association of specific biomarkers with immunotherapy survival (either PFS or OS) is assessed using Cox proportional hazards model to assess the discrimination or separation of a survival model. When needed p-values are corrected for multiple testing using the Benjamini-Hochberg (False Discovery Rate, FDR) method. Associations are deemed statistically significant for p-values and/or FDR lower or equal to 5\%. Note that Coef represents the log hazard ratio (logHR) or log odds ratio (logOR) depending on the analysis. 

## Association of Genes with Immunotherapy Overall Survival Response

We assess the association of protein-coding genes in Mariathasan__Bladder__PD-(L)1 with immunotherapy overall survival (OS), to gain more insight into the mechanism of response and resistance to immunotherapy response. 

```{r gene association os , message=FALSE}

## Identify the genes' association with overall survival (OS) in a dataset
## Analyze only protein-coding genes? (TODO: Clarify if needed)

# Define study details for reference in analysis
study_id <- "ICB_Ravi__Lung__PD-(L)1"
study_id_parts <- unlist(strsplit(study_id, "__"))

# Load expression and clinical data
#load(file.path(dir, paste(study_id, ".rda", sep="")))

# Apply a function over the first 1000 genes in the dataset.
cox_os <- lapply(1:100, function(i){
      
      # Call the geneSurvCont function to calculate the association of a single gene 
      # with overall survival using gene expression and clinical data.
      res <- geneSurvCont(dat.icb = expr,
                          clin = clin,
                          time.censor = 36,          # Censor patients after 36 months
                          missing.perc = 0.5,        # Exclude genes >50% missing data
                          const.int = 0.001,         # Small constant for internal calculations
                          n.cutoff = 15,             # Minimum number of samples required
                          feature = rownames(expr)[i], # i-th gene
                          study = paste(study_id_parts[1], study_id_parts[2], study_id_parts[3], sep="__"),
                          surv.outcome = "OS",      # Analyze overall survival
                          cancer.type = study_id_parts[2], # Cancer type
                          treatment = study_id_parts[3])   # Treatment type
       
      # Filter results without a coefficient value.
      res <- res[!is.na(res$Coef), ]
      return(res)
})

# Combine the analysis results into one data frame and exclude rows without a gene name.
cox_os <- do.call(rbind, cox_os)
cox_os <- cox_os[!is.na(cox_os$Gene), ]

# Adjust p-values for multiple testing with the Benjamini-Hochberg method to control the false discovery rate.
cox_os$FDR <- p.adjust(cox_os$Pval, method = "BH")
     
# Generate and display a formatted table of the top results, sorted by FDR.
knitr::kable(head(cox_os[order(cox_os$FDR, decreasing = FALSE), ]), format="html", 
             table.attr = "style='width:30%;'", row.names = FALSE) %>% 
             kableExtra::kable_styling()

     
```

The association results with $FDR < 0.05$ can be plotted as follows.     
     
```{r Fig1, message=FALSE, fig.width=5, fig.height=5, fig.cap = "genes association with immunotherapy overall survival", fig.align="center"}

p <- volcanoPlot(feature = cox_os$Gene,
                 coef = cox_os$Coef,
                 pval = cox_os$Pval,
                 padj = cox_os$FDR,
                 neg.cutoff = 11, # change based on your results
                 pos.cutoff = 1, # change based on your results
                 x.lab = "logHR estimate",
                 padj.label = FALSE, # change based on your results: if you have enough significant results based on FDR, then change that into TRUE
                 cutoff = 0.05)

p




     
```
     
Additionally, we assess the association of genes in PredictIO signature with immunotherapy response (R vs NR).
     
```{r gene association response, message=FALSE}
     
## identify genes' association with binary response (R vs NR)
## Note to run the following analysis for all the genes i.e., nrow(dat_expr)

# Apply a function over the first 200 genes in the dataset.
logreg <- lapply(1:nrow(expr), function(i){ 
  
      # Call the geneLogReg function which calculates the association of a single gene (i-th gene) 
      # with the response to treatment, using gene expression data (expr), clinical data (clin),
      # and various parameters for the analysis.
      res <- geneLogReg(dat.icb = expr,
                  clin = clin,
                  missing.perc = 0.5, # Exclude genes with more than 50% missing data
                  const.int = 0.001, # Small constant added to the data for internal reasons
                  n.cutoff = 15, # Minimum number of samples required
                  feature = rownames(expr)[i], # i-th gene
                  study = paste(strsplit("ICB_Wolf__Breast__IO+chemo", "__")[[1]][1], 
                                strsplit("ICB_Wolf__Breast__IO+chemo", "__")[[1]][2],
                                strsplit("ICB_Wolf__Breast__IO+chemo", "__")[[1]][3],
                                sep="__"), # Study ID created by concatenating parts of a string
                  n0.cutoff = 3, # Minimum number of samples for one group
                  n1.cutoff = 3, # Minimum number of samples for another group
                  cancer.type = strsplit("ICB_Wolf__Breast__IO+chemo", "__")[[1]][2], # Type of cancer
                  treatment = strsplit("ICB_Wolf__Breast__IO+chemo", "__")[[1]][3]) # Treatment type
  
      # Filter out the results that do not have a coefficient (Coef) value.
      res <- res[!is.na(res$Coef), ]
      
      # Return the result
      res
      
 })
     
# Combine the results of the analysis for each gene into one data frame.
logreg <- do.call(rbind, logreg)

# Exclude any rows that do not have a gene name.
logreg <- logreg[!is.na(logreg$Gene), ]

# Adjust the p-values for multiple testing using the Benjamini-Hochberg method to control the false discovery rate.
logreg$FDR <- p.adjust(logreg$Pval, method = "BH")

# Generate a formatted table of the top results, sorted by FDR in ascending order, 
# and display the top entries with HTML formatting for better visual presentation.
knitr::kable(head(logreg[order(logreg$FDR, decreasing = FALSE), ]), format="html", 
                  table.attr = "style='width:50%;'", row.names = FALSE) %>% kableExtra::kable_styling()
     
```
  
     
```{r Fig2, message=FALSE, fig.width=5, fig.height=4, fig.cap = "gene association with immunotherapy response (responder vs non-responder)", fig.align="center"}
     
p <- volcanoPlot(feature = logreg$Gene, 
                 coef = logreg$Coef, 
                 pval = logreg$Pval, 
                 padj = logreg$FDR, 
                 neg.cutoff = 10, # change based on your results 
                 pos.cutoff = 10, # change based on your results
                 x.lab = "logOR estimate",
                 padj.label = FALSE, # change based on your results
                 cutoff = 0.05)
     
p

# library(ggplot2)
# 
# # Create a new column to define the color of points
# logreg$color <- ifelse(logreg$Pval < 0.05 & logreg$Coef < 0, "blue",
#                        ifelse(logreg$Pval < 0.05 & logreg$Coef > 0, "purple", "grey"))
# 
# ggplot(logreg, aes(x = Coef, y = -log10(Pval), color = color)) +
#   geom_point() +
#   scale_color_identity() +
#   geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
#   theme_minimal() +
#   labs(title = "Volcano Plot", x = "logOR estimate", y = "-log10 P Value") +
#   geom_text(aes(label = ifelse(Pval < 0.05, as.character(Gene), '')),
#             hjust = 0.5, vjust = -0.5, size = 3, check_overlap = TRUE)
     
```

### Aggregating Associations (OS) through Meta-analysis (Pan-cancer)
     
To improve reproducibility, the results of individual independent datasets are pooled using random-effects meta-analysis with inverse variance weighting in \textit{DerSimonian and Laird} random-effects models. Heterogeneity across studies is evaluated by using the $Q$ statistic along with $I^2$ index, which describes the total variation across studies attributable to heterogeneity rather than sampling error. Note that $I^2$ value of $>50\%$ along with Cochranâ€™s $Q$ statistic $p < 0.05$ represents moderate-to-high heterogeneity. 

As an example, for a gene CXCL9, to generalize the association with immunotherapy survival, we apply a meta-analysis approach to integrate findings across datasets for pan-cancer, cancer-specific, and treatment-specific analyses. 

```{r CXCL9 OS}

## load SE object of IO data
## Note: load all new curated data (yours and mine)

# PDCD1, CD274, PDCD1LG2, CTLA4, CXCL9, F2RL1, CXCR6.

# # TODO: dir directory make sense to me here 
dir <- file.path(app_dir, 'PredictioR', "recent_data/se objects/")
# # gene expression, clinical characteristics, and gene meta data 
# load(file.path(dir, "ICB_IMmotion150__Kidney__IO+targeted.rda") )

# Load a list of files from a specified directory and apply a function to each file
expr <- lapply(1:length(list.files(dir)), function(k){ 
  load(file.path(dir, list.files(dir)[k])) 
  dat_icb
})

# Extract study names from filenames, assuming they start from the 5th character and end before '.rda'
study_icb <- substr(list.files(dir), 5, nchar(list.files(dir)) - 4)
# Assign these study names to the list elements in 'expr'
names(expr) <- study_icb

# Apply a function over the loaded datasets to perform survival analysis
cox_os <- lapply(1:length(expr), function(i){

# PDCD1, CD274, PDCD1LG2, CTLA4, CXCL9, F2RL1, CXCR6.
       # Call the geneSurvCont function for a specified gene 'CXCL9', using the dataset from 'expr'
       # and parameters such as time to censor, missing data percentage, etc.
       res <- geneSurvCont(dat.icb = expr[[i]],
                    time.censor = 36, # Censor time at 36 months
                    missing.perc = 0.5, # Exclude if more than 50% data is missing
                    const.int = 0.001, # A small constant to avoid division by zero or log(0)
                    n.cutoff = 15, # Minimum number of samples required
                    feature = "PDCD1LG2", # Gene of interest
                    study = paste(strsplit(names(expr)[i], "__")[[1]][1], 
                                  strsplit(names(expr)[i], "__")[[1]][2],
                                  strsplit(names(expr)[i], "__")[[1]][3],
                                  sep="__"), # Study ID created from the file name
                    surv.outcome = "OS", # Analyze overall survival
                    cancer.type = strsplit(names(expr)[i], "__")[[1]][2], # Cancer type from file name
                    treatment = strsplit(names(expr)[i], "__")[[1]][3]) # Treatment type from file name
       
       # Remove rows with NA coefficients from the results
       res <- res[!is.na(res$Coef), ]
       # Return the result
       res
})

# Combine results into one data frame
cox_os <- do.call(rbind, cox_os)

# Remove rows without a gene name
cox_os <- cox_os[!is.na(cox_os$Gene), ]

# Adjust p-values for multiple comparisons using the Benjamini-Hochberg method
cox_os$FDR <- p.adjust(cox_os$Pval, method = "BH")

# Create a HTML table with the top results sorted by the adjusted p-values
# Format the table width and apply additional styling from the kableExtra package
knitr::kable(head(cox_os[order(cox_os$FDR, decreasing = FALSE), ]), format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>%
                  kableExtra::kable_styling()



```
     
The integrated association shows CXCL9 is associated with worse survival with logHR = -0.22 and P < 0.001, where the heterogeneity across studies is approximately 34%.  

```{r meta-analysis pan-cancer}
     
# meta-analysis for a gene across datasets
res_meta <- meta.fun(coef = cox_os$Coef, 
                     se = cox_os$SE,
                     study  = cox_os$Study, 
                     pval = cox_os$Pval, 
                     n = cox_os$N,
                     cancer.type = cox_os$Cancer_type,
                     treatment = cox_os$Treatment,
                     feature = "PDCD1LG2", 
                     cancer.spec = FALSE, 
                     treatment.spec = FALSE)
     
# meta-analysis results
res_meta$meta_output
     
# summary of meta-analysis results
knitr::kable(res_meta$meta_summery, format="html", 
                  table.attr = "style='width:50%;'", row.names = FALSE) %>% kableExtra::kable_styling()
    
```
     
The following forest plot displays the aggregated results across datasets for pan-cancer analysis. For each dataset, the logHR, 95\% confidence intervals (CIs), and p-values are computed.  Horizontal bars represent the 95\% CIs of logHR. The blue diamond represents the aggregated results.
     
```{r Fig3, fig.width=8, fig.height=8, fig.cap ="Forest plot (CXCL9): Pan-cancer analysis", fig.align= 'center'}
     
forestPlot(coef = cox_os$Coef, 
           se = cox_os$SE, 
           study  = cox_os$Study, 
           pval = cox_os$Pval, 
           n = cox_os$N, 
           cancer.type = cox_os$Cancer_type,
           treatment = cox_os$Treatment,
           feature = "PDCD1LG2", 
           xlab = "logHR estimate", 
           label = "logHR")
     
```
     
### Aggregating Associations (response, R/NR) through Meta-analysis (Pan-cancer)

```{r CXCL9 response}

#365
# PDCD1, CD274, PDCD1LG2, CTLA4, CXCL9, F2RL1, CXCR6.

logreg <- lapply(1:length(expr), function(i){
    
      res <- geneLogReg(dat.icb = expr[[i]],
                  missing.perc = 0.5,
                  const.int = 0.001,
                  n.cutoff = 15,
                  feature = "CD274",
                  study = paste(strsplit(names(expr)[i], "__")[[1]][1], 
                                strsplit(names(expr)[i], "__")[[1]][2],
                                strsplit(names(expr)[i], "__")[[1]][3],
                                sep="__"), 
                  n0.cutoff = 3,
                  n1.cutoff = 3,
                  cancer.type = strsplit(names(expr)[i], "__")[[1]][2],
                  treatment = strsplit(names(expr)[i], "__")[[1]][3])
  
      res <- res[!is.na(res$Coef), ]
      
      res
      
 })
     
logreg <- do.call(rbind, logreg)
logreg <- logreg[!is.na(logreg$Gene), ]
logreg$FDR <- p.adjust(logreg$Pval, method = "BH")

knitr::kable(head(logreg[order(logreg$FDR, decreasing = FALSE), ]), format="html", 
                  table.attr = "style='width:50%;'", row.names = FALSE) %>% kableExtra::kable_styling()


```
     
Integrate the association between CXCL9 and response in cancer-specific setting. The results show that CXCL9 is negatively associated with response with logOR = -0.43 and P = 0.007, where there is zero heterogeneity across studies. 

     
```{r meta-analysis per-cancer}
     
# cancer specific meta-analysis for a gene across datasets
res_meta <- metaPerCan.fun(coef = logreg$Coef, 
                           se = logreg$SE,
                           study  = logreg$Study, 
                           pval = logreg$Pval, 
                           n = logreg$N,
                           cancer.type = logreg$Cancer_type,
                           treatment = logreg$Treatment,
                           feature = "PDCD1", 
                           cancer.spec = TRUE)
     
# summary of meta-analysis results for each cancer types
knitr::kable(rbind(res_meta$Kidney$meta_summery,
                   res_meta$Melanoma$meta_summery,
                   res_meta$Lung$meta_summery,
                   res_meta$Other$meta_summery), format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
    
```

The following forest plot displays the aggregated results across Melanoma datasets. For each dataset, the logHR, 95\% confidence intervals (CIs), and p-values are computed. Horizontal bars represent the 95\% CIs of logHR. The blue diamond represents the aggregated results.


Subgroup meta-analysis analysis is considered to assess the impact of cancer type on the source of moderate-to-high heterogeneity. 

```{r Fig4, fig.width=8, fig.height=12, fig.cap ="Forest plot (CXCL9): cancer-specific analysis", fig.align= 'center'}

forestPlotPerCan(coef = logreg$Coef, 
                 se = logreg$SE, 
                 study  = logreg$Study, 
                 pval = logreg$Pval, 
                 n = logreg$N, 
                 cancer.type = logreg$Cancer_type,
                 treatment = logreg$Treatment,
                 feature = "PDCD1", 
                 xlab = "logOR estimate", 
                 label = "logOR")
     
```


### Aggregating Associations (response, R/NR) through Meta-analysis (Per-treatment)

For treatment-specific analysis, consider meta-analysis when there are at least 3 datasets. The following forest plot displays the aggregated results across datasets as treatment-specific analysis. 

```{r meta-analysis per-treatment}
     
# meta-analysis for a gene across datasets
res_meta <- metaPerTreatment.fun(coef = logreg$Coef, 
                                 se = logreg$SE,
                                 study  = logreg$Study, 
                                 pval = logreg$Pval, 
                                 n = logreg$N,
                                 cancer.type = logreg$Cancer_type,
                                 treatment = logreg$Treatment,
                                 treatment.spec = TRUE,
                                 feature = "CD274")
     
# summary of meta-analysis results
knitr::kable(res_meta$`PD-(L)1`$meta_summery, format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
    
```

The following forest plot displays the aggregated results across PD-1/PD-L1 datasets. For each dataset, the logHR, 95\% confidence intervals (CIs), and p-values are computed. Horizontal bars represent the 95\% CIs of logHR. The blue diamond represents the aggregated results.

```{r Fig5a, fig.width=8, fig.height=6, fig.cap ="Forest plot (CXCL9, binary median cutoff): treatment-specific analysis", fig.align= 'center'}

dat <- res_meta$`PD-(L)1`$input_data   
forestPlot(coef = dat[ , "Coef"], 
           se = dat[ , "SE"], 
           study  = do.call(rbind, strsplit(dat[, "Study"], split=',', fixed=TRUE))[, 1], 
           pval = dat[ , "Pval"], 
           n = dat[ , "N"],
           cancer.type = dat[, "Cancer_type"],
           treatment = dat[, "Treatment"],
           feature = "CD274", 
           xlab = "logHR estimate", 
           label = "logHR")
     
```

```{r Fig5b, fig.width=8, fig.height=12, fig.cap ="Forest plot (CXCL9, binary median cutoff): treatment-specific analysis", fig.align= 'center'}

forestPlotPerTreatment(coef = logreg$Coef, 
                       se = logreg$SE, 
                       study  = logreg$Study, 
                       pval = logreg$Pval, 
                       n = logreg$N, 
                       cancer.type = logreg$Cancer_type,
                       treatment = logreg$Treatment,
                       feature = "CXCL9", 
                       xlab = "logOR estimate", 
                       label = "logOR")
     
```


